src/adapter/src/coord/ddl.rs

```diff
@@ -751,23 +736,10 @@ impl Coordinator {
         }
     }
 
-    pub(crate) fn drop_compute_sinks(&mut self, sinks: impl IntoIterator<Item = ComputeSinkId>) {
+    pub(crate) async fn drop_compute_sinks(&mut self, sinks: impl IntoIterator<Item = (ComputeSinkId, SubscribeRemovalReason)>) {
         let mut by_cluster: BTreeMap<_, Vec<_>> = BTreeMap::new();
-        for sink in sinks {
-            // Filter out sinks that are currently being dropped. When dropping a sink
-            // we send a request to compute to drop it, but don't actually remove it from
-            // `active_subscribes` until compute responds, hence the `dropping` flag.
-            //
-            // Note: Ideally we'd use .filter(...) on the iterator, but that would
-            // require simultaneously getting an immutable and mutable borrow of self.
-            let need_to_drop = self
-                .active_subscribes
-                .get(&sink.global_id)
-                .map(|meta| !meta.dropping)
-                .unwrap_or(false);
-            if !need_to_drop {
-                continue;
-            }
+        for (sink, reason) in sinks {
+            self.remove_active_subscribe(sink.global_id, reason).await;
 
             if !self
                 .controller
```

Comment:
@teskje there's now a code path where the compute read policy is not removed. If a `SUBSCRIBE ... UP TO` completes, then it will be removed from the adapter state when the `SubscribeResponse::DroppedAt` is processed... but `drop_compute_sinks` will thus never get called on that subscribe. That seems like a problem, and it's related to the next comment...

Reply:
https://github.com/MaterializeInc/materialize/pull/25179#discussion_r1487243233